docs/design_proposals/sync-auto-support.md

```diff
@@ -0,0 +1,150 @@
+    docker build and sync files from a built container to a running container --
+    something we do not intend to cover here.
+
+#### Problems with current API/config
+The current `sync` system has the following problems:
+1. No way to trigger local external processes - for example, in jib, to build a
+   container, one would run `./gradlew jib`, but to update class files so the
+   system may sync them, one would only be required to run `./gradlew classes`.
+2. No way to tell the system to sync non build inputs. Skaffold is
+   normally only watching `.java` files for a build, in the sync case, we want
+   it to watch `.java` files, trigger a partial build, and sync `.class` files
+   so a remote server can pick up and reload the changes.
+
+## Design
+
+#### Hack it
+To get close to the functionality we want, without modifying skaffold at all, a
+Dockerfile which depends on java build outputs could be used, like:
+```
+FROM openjdk:8
+COPY build/dependencies/ /app/dependencies
+COPY build/classes/java/main/ /app/classes
+
+CMD ["java", "-cp", "/app/classes:/app/dependencies/*", "hello.Application"]
+```
+
+with a skaffold sync block that looks like:
+```
+sync:
+  manual:
+    - src: "build/classes/java/main/**/*.class"
+      dest: "/app/classes"
+      strip: "build/classes/java/main/"
+```
+
+A user's devloop then looks like this:
+
+1. run `./gradlew classes copyDependencies`
+2. run `skaffold dev`
+3. *make changes to some java file*
+4. run `./gradlew classes`
+5. *skaffold syncs files*
+
+which is far from ideal.
+
+#### Fix it?
+
+What we might want, is an API that can accept configuration that looks like:
+
+```
+sync: auto {
+  indirect:
+  - command: "./gradlew jib classes"
+    inputs:
+      - src/main/java
+      - src/main/resources
+    files:
+      - src: "target/classes"
+        dest: "/app/classes",
+      - src: "target/resources.
+        dest: "/app/resources",
+  - command: "compile_c_stuff.sh"
+    inputs:
+    - asdf.c
+    files:
+     - src: "asdf.o"
+       dest: "/libs/asdf.o"
+  direct:
+    - src: "local/computer/extrafiles"
+      dest: "/extrafiles"
+  includeTools: true
+}
+```
+
+Lets breakdown the different parts
+
+1. `sync: auto` - just a name for the mode, can be anything
+2. `indirect` - this means, you can't sync files directly, you need a command to
+   tranform them before sync. Indirect is also a list that can handle multiple
+   indirect types.
+     1. `command` - the command to run to generate syncable files
+     2. `inputs` - the inputs to watch to trigger `command`, if left empty, this
+        should just be considered the full list of inputs that are normally
+        watched for this build minus anything in `direct` (below).
+     3. `files` - the built files to sync (and where to sync them)
+3. `direct` - same as manual sync, just copy them directly, don't wait on some
+   command to transform them
+4. `includeTools` - useful for containers that don't contain, for example, `tar`
```

Comment:
would this for cases where a generated file is bundled up in the container?
So the sync will now, 
- unzip the tar
- sync file 
- zip again?

Reply:
how do specify which tools to include? Is there a separate DSL for that?

Reply:
I think `includeTools` maybe shouldn't be part of this design. While it's useful specifically for the default base image jib chooses (distroless/java), it could benefit be it's own design that answers these questions.

But to answer them anyway, since I'm here, I think for now it would include `tar` -- since that's what we need to do a sync. If the requirements of sync change, it could include whatever else `sync` needs. (that's why naming would be `includeTools` instead of `includeTar`)

Reply:
Having an option to add `tar` into the target container is a great idea. I also think that this is an idea worth being explored on its own. For example, it could solve https://github.com/GoogleContainerTools/skaffold/issues/1814. Consequently, I would see this option outside of the `auto` block and right under `sync`, so that other sync modes can also benefit from it.

Reply:
hmm. i am still confused on it makes sense for me as a user writing this config.
```
sync:
   inlcudeTar: true
  generated:
   ....
```
If this is true,  skaffold would now install tar inside all containers built ?
If this is false, skaffold sync will fail potentially #1814 ?

If above is true; then 
As a user, if i have a  jib builder, i need to set this `true`.

I feel the users are might not be able to identify this. Can we instead, add an interface to builder
```
type Builder interface {
   SyncTools() []tools
}
```
And then skaffold in dev mode if sync enabled, will install these tools in the built container?

Reply:
It would make tar available to that specific artifacts container. Not all jib users *have* to do this, but the `distroless` collection of images do not have tar (or any tools really) included by default.

I would think, instead of asking the user to change their base image to enable sync, we just make it easier for them by making the tools available automatically. I assume it would a mechanism similar to how debugging tools are available to a pod right now using init containers (@briandealwis ?).

Again, perhaps I should take this out of this design doc and move it to a new one?

Reply:
Perhaps for simplicity, this doc could assume `tar` exists in the container or declares that this works only when `tar` exists. Installing other builds tools could be another doc, independent of this?

Reply:
ah understood, so maybe skaffold can be a little smart about if `tar` is included in the built image. 
Untill then, asking users to specify via config is easy way.
Skaffold can also suggest adding it to their config if the error is regarding `tar` not found.